### PR TITLE
feat: add k3d local devnet scripts

### PR DESCRIPTION
**Pull request by glebiller**
Add a set of kubernetes manifests to locally start a devnet containing:
* one genesis bootstrapping the chain and acting as a validator
* one seed node having the genesis as private persistent peer
* N (3 by default) validators bootstrap from the genesis node

For the moment it's using a manually deployed docker image of elestod on Google Container Registry.
It would be good to switch to GitHub Container Registry as soon as we start deploying there :)

### CODE REVIEW

File: scripts/devnet/kubernetes/0-init/init-genesis.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+fi
+
+# Create key if it does not exists
+if [ "$(elestod keys list --output json --home /home/ | jq -r --arg MONIKER "${MONIKER}" '.[] | select(.name == $MONIKER).name')" != "${MONIKER}" ]; then
+  echo "Create ${MONIKER} key"
+  elestod keys add ${MONIKER} --home=/home/ --output json | jq -r '.mnemonic' > /home/config/${MONIKER}_mnemonic
+  elestod add-genesis-account $(elestod keys show ${MONIKER} --home=/home/ -a) ${GENESIS_AMOUNT}stake --home=/home/
+fi
+
+# Create regulator
```
Comment (barriebyron) [inline scripts/devnet/kubernetes/0-init/init-genesis.sh]:
```suggestion
# Create regulator key
```

File: scripts/devnet/kubernetes/0-init/init-genesis.sh [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+if [ "$(elestod keys list --output json --home /home/ | jq -r '.[] | select(.name == "regulator").name')" != "regulator" ]; then
+  echo "Create regulator key"
+  echo ${REGULATOR_KEY} | elestod keys add regulator --home=/home/ --output json --recover
+  elestod add-genesis-account $(elestod keys show regulator --home=/home/ -a) 20000000stake --home=/home/
+fi
+
+# Create emti (e-money token issuer)
```
Comment (barriebyron) [inline scripts/devnet/kubernetes/0-init/init-genesis.sh]:
```suggestion
# Create e-money token issuer (EMTI) key
```

Comment (barriebyron) [reply scripts/devnet/kubernetes/0-init/init-genesis.sh]: I'm curious, is this key for the e-money token itself? or a key to be used by the issuer only?

File: scripts/devnet/kubernetes/0-init/init-genesis.sh [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+if [ "$(elestod keys list --output json --home /home/ | jq -r '.[] | select(.name == "emti").name')" != "emti" ]; then
+  echo "Create emti key"
+  echo ${EMTI_KEY} | elestod keys add emti --home=/home/ --output json --recover
+  elestod add-genesis-account $(elestod keys show emti --home=/home/ -a) 20000000stake --home=/home/
+fi
+
+# Create arti (asset-referenced token issuer)
```
Comment (barriebyron) [inline scripts/devnet/kubernetes/0-init/init-genesis.sh]:
```suggestion
# Create asset-referenced issuer (ARTI) token
```

File: scripts/devnet/kubernetes/0-init/init-genesis.sh [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  echo "Generate transaction and collect to Genesis on chain ${CHAIN_ID}"
+  elestod gentx ${MONIKER} 70000000stake --home=/home/ --chain-id=${CHAIN_ID}
+  elestod collect-gentxs --home=/home/
+fi
+
+# Update Genesis
+jq '.app_state["staking"]["params"]["unbonding_time"] = "240s"' /home/config/genesis.json > /home/config/genesis.json.tmp  && mv /home/config/genesis.json.tmp /home/config/genesis.json
+jq '.app_state["gov"]["voting_params"]["voting_period"] = "60s"' /home/config/genesis.json > /home/config/genesis.json.tmp  && mv /home/config/genesis.json.tmp /home/config/genesis.json
+
+# Create/Update genesis configmap
```
Comment (barriebyron) [inline scripts/devnet/kubernetes/0-init/init-genesis.sh]:
```suggestion
# Create or update genesis configmap
```

File: scripts/devnet/kubernetes/0-init/init-validator.sh [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  sed -i 's|keyring-backend = ".*"|keyring-backend = "test"|' /home/config/client.toml
+fi
+
+# Add ${INIT_FROM} private key
+if [ "$(elestod keys list --output json --home /home/ | jq -r --arg INIT_FROM "${INIT_FROM}" '.[] | select(.name == $INIT_FROM).name')" != "${INIT_FROM}" ]; then
+  echo "Add ${INIT_FROM} key"
+  kubectl get secret ${INIT_FROM} --output json | jq -r '.data.MNEMONIC' | base64 -d | elestod keys add ${INIT_FROM} --home=/home/ --output json --recover
+fi
+
+# Create key if it does not exists
```
Comment (barriebyron) [inline scripts/devnet/kubernetes/0-init/init-validator.sh]:
```suggestion
# Create a key if it doesn't exist
```

File: scripts/devnet/kubernetes/0-init/init-validator.sh [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  elestod tx bank send ${FROM_ADDRESS} ${TO_ADDRESS} ${AMOUNT}stake --home=/home/ --chain-id=${CHAIN_ID} --from ${FROM_ADDRESS} --node tcp://${INIT_FROM%-*}:26657 --output json --yes
+
+  # Wait for transaction to complete & check
+  while [ "$(elestod query bank balances ${TO_ADDRESS} --home=/home/ --chain-id=${CHAIN_ID} --node tcp://${INIT_FROM%-*}:26657 --output json | jq -r '.balances[] | select(.denom="stake").amount')" != "${AMOUNT}" ]; do
+    echo "tx bank send transaction still not processed (amount not found)"
+    sleep 1;
+  done
+fi
+
+# Create validator
```
Comment (barriebyron) [inline scripts/devnet/kubernetes/0-init/init-validator.sh]:
```suggestion
# Create or register the validator node on the chain
```

Comment (barriebyron) [reply scripts/devnet/kubernetes/0-init/init-validator.sh]: what kind of a validator are we creating here? a validator account? (node?) teach me!

File: scripts/devnet/kubernetes/README.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# Local blockchain on Kubernetes
```
Comment (barriebyron) [inline scripts/devnet/kubernetes/README.md]:
```suggestion
# Create a local blockchain on a Kubernetes cluster

Start a local devnet with:

- One genesis to bootstrap the chain and act as a validator
- One seed node with the genesis as a private persistent peer
- N (3 by default) validators that bootstrap from the genesis node
```

File: scripts/devnet/kubernetes/README.md [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+## Using k3d
+
+`make k3d-start k3d-deploy`
+
+## Accessing a node
+
+`make port-forward` or `kubectl port-forward --namespace local svc/genesis 26657`
+Then `curl localhost:26657/validators | jq '.result.validators[].address'`
+
+## ER Diagram
```
Comment (barriebyron) [inline scripts/devnet/kubernetes/README.md]:
```suggestion
## Entity relationship diagram
```

Comment (barriebyron) [reply scripts/devnet/kubernetes/README.md]: btw, this mermaid is so cool

**Review by barriebyron — COMMENTED** [review]
@glebiller  mostly I review to learn and expand my knowledge... I contributed a few grammatical suggestions for human-readable text

Love the mermaid diagram to show the entity relationship

Lmk if this type of feedback and questions are useful

**Comment by gsora** [comment]
Can I test this in some way?

I think I need to be added to some user group:

```
 => ERROR [internal] load metadata for us-central1-docker.pkg.dev/elesto/elesto/elestod:2.0.0-rc2-readiness                                                                                                               1.4s
------
 > [internal] load metadata for us-central1-docker.pkg.dev/elesto/elesto/elestod:2.0.0-rc2-readiness:
------
failed to solve with frontend dockerfile.v0: failed to create LLB definition: failed to authorize: rpc error: code = Unknown desc = failed to fetch anonymous token: unexpected status: 403 Forbidden
```

File: scripts/devnet/kubernetes/0-init/init-validator.sh (hunk 6 above)
Comment (gsora) [reply scripts/devnet/kubernetes/0-init/init-validator.sh]: In this instance the script is creating a validator node, which *does* have a validator account :)

Comment (glebiller) [reply scripts/devnet/kubernetes/0-init/init-validator.sh]: I will keep it as "Create the validator", as the account is created during the "Create a key" phase. Here we are really creating or registering the validator node on the blockchain.

File: scripts/devnet/kubernetes/README.md (hunk 8 above)
Comment (glebiller) [reply scripts/devnet/kubernetes/README.md]: Thanks!

File: scripts/devnet/kubernetes/0-init/init-validator.sh (hunk 6 above)
Comment (barriebyron) [reply scripts/devnet/kubernetes/0-init/init-validator.sh]: that description is perfect! let's be clear with our words to describe the action

**Comment by glebiller** [comment]
I commit the new version that uses the ghcr.io docker image instead of building & importing them in k3d.
The only thing is that since we are still using private repositories, we still need to create a registries.yaml file that contains a GitHub PAT so k3d can download the containers.

**Review by gsora — APPROVED** [review]

**Pull request merged**